src/core/providers/ept/qgseptprovider.cpp

```diff
@@ -73,29 +86,66 @@ QString QgsEptProvider::description() const
   return QStringLiteral( "Point Clouds EPT" );
 }
 
-QgsPointCloudIndex *QgsEptProvider::index()
+QgsPointCloudIndex *QgsEptProvider::index() const
 {
   return mIndex.get();
 }
 
 int QgsEptProvider::pointCount() const
 {
-  return mIndex->pointCount();
+  if ( mIndex )
+    return mIndex->pointCount();
+  else
+    return 0;
 }
 
 QVariantList QgsEptProvider::metadataClasses( const QString &attribute ) const
 {
-  return mIndex->metadataClasses( attribute );
+  if ( mIndex )
+    return mIndex->metadataClasses( attribute );
+  else
+    return QVariantList();
 }
 
 QVariant QgsEptProvider::metadataClassStatistic( const QString &attribute, const QVariant &value, QgsStatisticalSummary::Statistic statistic ) const
 {
-  return mIndex->metadataClassStatistic( attribute, value, statistic );
+  if ( mIndex )
+    return mIndex->metadataClassStatistic( attribute, value, statistic );
+  else
+    return QVariant();
+}
+
+void QgsEptProvider::loadIndex()
+{
+  if ( mRunningIndexingTask || mIndex )
+    return;
+
+  mRunningIndexingTask = nullptr;
+  QgsEptPointCloudIndexLoadingTask *task = new QgsEptPointCloudIndexLoadingTask( mUri );
+  connect( task, &QgsEptPointCloudIndexLoadingTask::taskTerminated, this, &QgsEptProvider::onLoadIndexFinished );
+  connect( task, &QgsEptPointCloudIndexLoadingTask::taskCompleted, this, &QgsEptProvider::onLoadIndexFinished );
+  mRunningIndexingTask = task;
+  QgsApplication::taskManager()->addTask( task );
+}
+
+void QgsEptProvider::onLoadIndexFinished()
+{
+  QgsEptPointCloudIndexLoadingTask *task = qobject_cast<QgsEptPointCloudIndexLoadingTask *>( QObject::sender() );
+  // this may be already canceled task that we don't care anymore...
+  if ( task == mRunningIndexingTask )
+  {
+    mIndex = mRunningIndexingTask->index();
```

Comment:
This should probably be something like "takeIndex" which transfers ownership

Reply:
That said - QgsPointCloudIndex is currently a qobject, so we have to be careful moving it between threads. I can't see any reason it HAS to be a qobject though, so I suggest we just remove that...

Reply:
ok, talked with Martin and the ept load will be in main thread for now (should be quick from files)